### PR TITLE
Fix null CacheManager when getting CacheManager concurrently

### PR DESCRIPTION
**Pull request by zzcclp**
### What changes are proposed in this pull request?
Use `CACHE_INIT_LOCK.lock()` instead of `CACHE_INIT_LOCK.tryLock()` when  calling 'CacheManager.Factory.get(conf)' concurrently, to make sure it gets non-null CacheManager.

This PR fixes https://github.com/Alluxio/alluxio/issues/13640

### Why are the changes needed?
Returning null when calling 'CacheManager.Factory.get(conf)' concurrently.

I used Spark 3.1.1 + Alluxio Local Cache, when all tasks in one executor start concurrently, there are some tasks getting null CacheManager.

### Does this PR introduce any user facing changes?
No

### CODE REVIEW

File: core/client/fs/src/main/java/alluxio/client/file/cache/CacheManager.java [hunk 1]
```diff
@@ -84,17 +84,16 @@ public int getValue() {
...
-            if (CACHE_MANAGER.get() == null) {
-              CACHE_MANAGER.set(create(conf));
-            }
-          } catch (IOException e) {
-            Metrics.CREATE_ERRORS.inc();
-            throw new IOException("Failed to create CacheManager", e);
-          } finally {
-            CACHE_INIT_LOCK.unlock();
+        CACHE_INIT_LOCK.lock();
+        try {
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/file/cache/CacheManager.java:88]:
```suggestion
        try (LockResource r = new LockResource(CACHE_INIT_LOCK)) {

```

File: core/client/fs/src/main/java/alluxio/client/file/cache/CacheManager.java [hunk 2]
```diff
@@ -84,17 +84,16 @@ public int getValue() {
...
+          if (CACHE_MANAGER.get() == null) {
+            CACHE_MANAGER.set(create(conf));
           }
+        } catch (IOException e) {
+          Metrics.CREATE_ERRORS.inc();
+          throw new IOException("Failed to create CacheManager", e);
+        } finally {
+          CACHE_INIT_LOCK.unlock();
         }
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/file/cache/CacheManager.java:97]: With `LockResource` we don't need `finally`
```suggestion
        }
```

File: core/client/fs/src/test/java/alluxio/client/file/cache/CacheManagerTest.java [hunk 3]
```diff
@@ -40,4 +62,21 @@ public void factoryClear() throws Exception {
...
+    @Override
+    public CacheManager call() throws Exception {
+      try {
+        return CacheManager.Factory.get(mConf);
+      } catch (Exception e) {
+        return null;
+      }
+    }
+  }
+}
```
Comment (apc999) [inline core/client/fs/src/test/java/alluxio/client/file/cache/CacheManagerTest.java:82]: a new line please

File: core/client/fs/src/test/java/alluxio/client/file/cache/CacheManagerTest.java [hunk 4]
```diff
@@ -40,4 +62,21 @@ public void factoryClear() throws Exception {
...
     assertNotEquals(manager, CacheManager.Factory.get(mConf));
   }
-}
+
+  private static class CacheManagerFetch implements Callable<CacheManager> {
+    InstancedConfiguration mConf = null;
+
+    public CacheManagerFetch(InstancedConfiguration mConf) {
+      this.mConf = mConf;
+    }
```
Comment (apc999) [inline core/client/fs/src/test/java/alluxio/client/file/cache/CacheManagerTest.java:71]: Alluxio code convention:

```suggestion
    public CacheManagerFetch(InstancedConfiguration conf) {
      mConf = mConf;
    }
```

File: core/client/fs/src/test/java/alluxio/client/file/cache/CacheManagerTest.java [hunk 5]
```diff
@@ -28,6 +36,20 @@ public void factoryGet() throws Exception {
     assertEquals(manager, CacheManager.Factory.get(mConf));
   }
 
+  @Test
+  public void factoryGetConcurrently() throws Exception {
```
Comment (apc999) [inline core/client/fs/src/test/java/alluxio/client/file/cache/CacheManagerTest.java:40]: shall we call `CacheManager.clear()` first?

File: core/client/fs/src/test/java/alluxio/client/file/cache/CacheManagerTest.java [hunk 6]
```diff
@@ -28,6 +36,20 @@ public void factoryGet() throws Exception {
...
+    int cnt = 10;
+    List<Future<CacheManager>> cacheManagersList = new ArrayList<>(cnt);
+    ExecutorService executorService = Executors.newFixedThreadPool(cnt);
+    for (int i = 0; i < cnt; i++) {
+      cacheManagersList.add(executorService.submit(new CacheManagerFetch(mConf)));
+    }
+
+    for (Future<CacheManager> f : cacheManagersList) {
+      assertNotNull(f.get());
```
Comment (apc999) [inline core/client/fs/src/test/java/alluxio/client/file/cache/CacheManagerTest.java:49]: and they shall all be the same handle, right?

**Review by apc999 — CHANGES_REQUESTED** [review]
Comments left. Thanks for the improvement!

File: core/client/fs/src/main/java/alluxio/client/file/cache/CacheManager.java (hunk 1 above)
Comment (zzcclp) [reply core/client/fs/src/main/java/alluxio/client/file/cache/CacheManager.java:88]: Done

File: core/client/fs/src/main/java/alluxio/client/file/cache/CacheManager.java (hunk 2 above)
Comment (zzcclp) [reply core/client/fs/src/main/java/alluxio/client/file/cache/CacheManager.java:97]: Done

**Comment by zzcclp** [comment]
> Comments left. Thanks for the improvement!

Already fixed.

File: core/client/fs/src/test/java/alluxio/client/file/cache/CacheManagerTest.java [hunk 7]
```diff
@@ -28,6 +36,24 @@ public void factoryGet() throws Exception {
...
   }
 
+  @Test
+  public void factoryGetConcurrently() throws Exception {
+    CacheManager.Factory.clear();
+    int cnt = 10;
+    List<Future<CacheManager>> cacheManagersList = new ArrayList<>(cnt);
+    ExecutorService executorService = Executors.newFixedThreadPool(cnt);
+    for (int i = 0; i < cnt; i++) {
+      cacheManagersList.add(executorService.submit(new CacheManagerFetch(mConf)));
```
Comment (apc999) [inline core/client/fs/src/test/java/alluxio/client/file/cache/CacheManagerTest.java:46]: nit: you can replace this with lambda expression for readability, without introducing a new class `CacheManagerFetch`

```suggestion
      cacheManagersList.add(executorService.submit(() -> {
        try {
          return CacheManager.Factory.get(mConf);
        } catch (Exception e) {
          return null;
        }
      }));
```

**Review by apc999 — CHANGES_REQUESTED** [review]
Two minor improvements suggested. Can you address them? 

Thanks

File: core/client/fs/src/main/java/alluxio/client/file/cache/CacheManager.java [hunk 8]
```diff
@@ -85,16 +85,14 @@ public int getValue() {
     public static CacheManager get(AlluxioConfiguration conf) throws IOException {
       // TODO(feng): support multiple cache managers
       if (CACHE_MANAGER.get() == null) {
-        if (CACHE_INIT_LOCK.tryLock()) {
+        try (LockResource lockResource = new LockResource(CACHE_INIT_LOCK)) {
           try {
```
[the remark this answers is not in the document]
Comment (zzcclp) [reply core/client/fs/src/main/java/alluxio/client/file/cache/CacheManager.java:89]: It needs, the `creating CacheManager` method throws IOException and it needs to catch.

File: core/client/fs/src/test/java/alluxio/client/file/cache/CacheManagerTest.java (hunk 7 above)
Comment (zzcclp) [reply core/client/fs/src/test/java/alluxio/client/file/cache/CacheManagerTest.java]: Done

File: core/client/fs/src/main/java/alluxio/client/file/cache/CacheManager.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (apc999) [reply core/client/fs/src/main/java/alluxio/client/file/cache/CacheManager.java:89]:
```
try (LockResource lockResource = new LockResource(CACHE_INIT_LOCK)) {
  if (CACHE_MANAGER.get() == null) {
    CACHE_MANAGER.set(create(conf));            
  } 
} catch (IOException e) {
  Metrics.CREATE_ERRORS.inc();
  throw new IOException("Failed to create CacheManager", e);
}
```
```

[the remark this answers is not in the document]
Comment (zzcclp) [reply core/client/fs/src/main/java/alluxio/client/file/cache/CacheManager.java]: Done

**Review by apc999 — APPROVED** [review]

**Comment by zzcclp** [comment]
retest this please

**Comment by apc999** [comment]
alluxio-bot, merge this please

**Pull request merged**